Add text iteration over children

Ygg01 opened the pull request:
By calling `node.text_iter()` you get `Ref<'a, String>` of all children that are text nodes. Using this Iterator allows you to bypass the `as_text` cast and `is_text` check, but it requires you to iterate over content.

Should I add a mutable version that returns `RefCell`?

SimonSapin commented on src/tree.rs:
```diff
@@ -158,6 +158,8 @@ impl<'a> Node<'a> {
         }
     }
 
+    pub fn is_text(&self) -> bool { matches!(self.data, NodeData::Text(_))}
```
Is this really useful, compared to `node.as_text().is_some()`? It also encourages redundant tests (see comment on the new iterator.)

SimonSapin commented on src/select.rs:
```diff
@@ -121,3 +130,18 @@ impl<'a,T> Iterator for FilterNodes<T> where T: Iterator<Item=&'a Node<'a>> {
+    fn next(&mut self) -> Option<Ref<'a, String>> {
+        for node in self.iter.by_ref() {
+            if node.is_text() {
+                return Some(node.as_text().unwrap().borrow())
```
Please avoid `.unwrap()` and use `if let` instead. (This also makes `is_text` unnecessary.)

SimonSapin commented on src/select.rs:
```diff
@@ -93,21 +95,28 @@ impl<'a> TElement<'a> for &'a ElementData {
+
+    pub fn text_iter(&'a self) -> FilterNodes<Siblings<'a>, TextFilter> {
+        FilterNodes {
+            iter: self.children(),
```
Is taking the text on only direct child text nodes (ignoring other descendants) ever the desired behavior?

SimonSapin commented on src/select.rs:
```diff
@@ -93,21 +95,28 @@ impl<'a> TElement<'a> for &'a ElementData {
         })
     }
+
+    pub fn text_iter(&'a self) -> FilterNodes<Siblings<'a>, TextFilter> {
```
Please don’t hijack `FilterNodes` for this. You’re already adding a new struct and a new `Iterator` impl. Just make that struct the iterator. (Feel free to rename `FilterNodes` to something more specific, though.)

Ygg01 replied:
Ok, so perhaps rename `FilterNodes` -> `SelectorNodes` and add `TextSelector`. 

When I look at `FilterNodes` I think there is some overlap with [NodeFilter](http://www.w3.org/TR/DOM-Level-2-Traversal-Range/traversal.html#Traversal-NodeFilter). `TextSelector` strucks me as a convenience of that filter.

SimonSapin commented:
When I first skimmed this yesterday I assumed it was text iteration over descendants. What’s the use case for getting child nodes and ignoring other descendants?

> Should I add a mutable version that returns `RefCell`?

Do you mean `RefMut`? That sounds fine.

Ygg01 commented:
> When I first skimmed this yesterday I assumed it was text iteration over descendants.

I meant this as a way to simulate the `.tail`/`.text` from `eTree`, those do not look at descendants as far as I can tell. I suppose descendants makes perhaps more sense. Or leave it configurable?

> Do you mean `RefMut`?

Well, I initially made a version that returned `RefCell`. Out of curiosity, what would be different if I returned `RefCell` instead of `RefMut`? Won't both of them panic on concurrent mutation?

SimonSapin commented:
I don’t really like the `.tail` / `.text` model, it was deliberate to have text nodes be "first class" nodes in Kuchiki. But regardless, even in that model, this is equivalent to taking the `.text` of an element and the `.tail` of all of its child elements (but again, ignoring other descendants and `.text` of children). This seems like an awfully specific and arbitrary algorithm. In what kind of case is this what you want?

To return a `RefCell` from a `&Node` you’d have to clone it. Or did you mean `&RefCell`? That would work, but I’d do that instead of rather than in addition to returning `Ref`.

Ygg01 commented:
Ok. Sounds good.

No, you're right. `RefMut` shows more intent. So I'll add `text_iter_mut()` which iterates over `RefMut`, sounds ok?

SimonSapin commented:
If we go with returning `&RefCell`, this new iterator could be entirely replaced with `.filter_map(Node::as_text)`.

Ygg01 committed:
Rename FilterNodes to SelectNodes

- As per our discussion, there is not reason for them to be called
FilterNodes.

Ygg01 committed:
Add text_iter method, which returns FilterMap

- Add a method that iterates over any descendant node that is text
- Method always returns a RefCell

Ygg01 committed:
Add tests for access and modification via Iterator

Ygg01 commented:
@SimonSapin Ok, I've done as you've asked. No more Iterator, it just returns a FilterMap. I did manage to clobber up [some lifetimes](https://github.com/Ygg01/kuchiki/commit/cb437c6c5384d9dccc59dc964e7c8c785f7a6ce7#diff-705fb728fe7444d7a90bdd8bde5447b9R108), so please check that :worried:

SimonSapin commented:
Looks good, thanks!

I don’t have a strong opinion on one iterator yielding `&RefCell` v.s. two yielding `Ref` and `RefMut`, so feel free to make another PR if the latter turns out to be more convenient.

tabatkins commented:
> I don’t really like the .tail / .text model, it was deliberate to have text nodes be "first class" nodes in Kuchiki. 

The .tail/.text model is indeed absolutely terrible in my experience.  I have a bunch of DOM-inspired functions wrapping all my insertion/etc so I can pretend that text nodes are first-class, as it makes _way_ more sense.

(The LXML text/tail model is specialized for XML languages that don't mix text and content; all elements either have child nodes or text or neither, never a mix of child nodes and text.  .text/.tail support that model well, while paying super-inconvenient lip service to mixed-content stuff like HTML.)
